include/ccf/ds/pal.h

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include <cstdint>
+#include <cstdlib>
+
+#if !defined(INSIDE_ENCLAVE) || defined(VIRTUAL_ENCLAVE)
+#  include <cstring>
+#  include <mutex>
+#else
+#  include <openenclave/attestation/attester.h>
+#  include <openenclave/bits/defs.h>
+#  include <openenclave/bits/security.h>
+#  include <openenclave/edger8r/enclave.h> // For oe_lfence
+#  include <openenclave/enclave.h>
+#  include <openenclave/log.h>
+#  include <openenclave/tracee.h>
+#  include <pthread.h>
+#endif
+
+/**
+ * This file implements a platform abstraction layer to enable platforms, such
+ * as OpenEnclave to offer custom implementations for certain functionalities.
+ * By centralizing the platform-specific code to one file, we can avoid exposing
+ * platform-specific types to the rest of the code and have a good overview of
+ * all the functionality that is custom to a given platform. The platform
+ * abstraction layer can also be used in code shared between the host and the
+ * enclave as there is a host implementation for it is well.
+ */
+namespace ccf
+{
+  /**
+   * Malloc information formatted based on the OE type, but avoiding to expose
+   * the actual OE type in non-OE code.
+   */
+  struct MallocInfo
+  {
+    size_t max_total_heap_size = 0;
+    size_t current_allocated_heap_size = 0;
+    size_t peak_allocated_heap_size = 0;
+  };
+
+#if !defined(INSIDE_ENCLAVE) || defined(VIRTUAL_ENCLAVE)
+  /**
+   * Virtual enclaves and the host code share the same PAL.
+   * This PAL takes no dependence on OpenEnclave, but also does not apply
+   * security hardening.
+   */
+  class HostPal
+  {
+  public:
+    using Mutex = std::mutex;
+
+    static inline void* safe_memcpy(void* dest, const void* src, size_t count)
+    {
+      return ::memcpy(dest, src, count);
+    }
+
+    static inline void speculation_barrier() {}
+
+    static inline void redirect_platform_logging() {}
+
+    static inline void initialize_enclave() {}
+
+    static inline void shutdown_enclave() {}
+
+    static inline bool is_outside_enclave(const void* ptr, std::size_t size)
+    {
+      return true;
+    }
+
+    static inline bool get_mallinfo(MallocInfo& info)
+    {
+      info.max_total_heap_size = std::numeric_limits<size_t>::max();
+      info.current_allocated_heap_size = 0;
+      info.peak_allocated_heap_size = 0;
+      return true;
+    }
+  };
+
+  using Pal = HostPal;
+
+#else
+  class OEPal
+  {
+    class MutexImpl
```

Comment:
Can you add a comment to clarify this exists because of https://github.com/openenclave/openenclave/issues/4555, and can be removed once a release of Open Enclave containing the fix is published?

Reply:
Thanks, updated!